### PR TITLE
Phase 5: NAT Generator Implementation

### PR DESCRIPTION
**Pull request by duplico**
## Summary

Implements Phase 5 of the vyos-onecontext implementation plan: NAT (Network Address Translation) configuration generation for VyOS Sagitta.

This PR adds comprehensive NAT support for three types of NAT configurations:

1. **Source NAT (SNAT/Masquerade)** - Outbound traffic translation
   - Dynamic masquerading (typical for internet gateway scenarios)
   - Static IP translation
   - Optional source address filtering

2. **Destination NAT (DNAT/Port Forwarding)** - Inbound traffic translation
   - Protocol filtering (tcp, udp, tcp_udp, icmp, or all)
   - Port translation/forwarding
   - Optional destination address filtering for 1:1 NAT scenarios

3. **Bidirectional 1:1 NAT** - Full bidirectional translation
   - Creates paired source and destination NAT rules
   - Designed for exposing internal servers with dedicated public IPs
   - Commonly used with OpenNebula NIC aliases for scoring engines

### New Files
- `src/vyos_onecontext/generators/nat.py` - Complete NAT generator implementation (234 lines)
  - `NatGenerator` class with separate methods for each NAT type
  - Rule numbering: regular NAT (100-400), binat (500+)
  - Full type annotations for mypy strict mode

### Modified Files
- `src/vyos_onecontext/generators/__init__.py` - Register NAT generator in pipeline
- `tests/test_generators.py` - 412 lines of comprehensive tests (41 test cases)

### Rule Numbering Strategy
- **Source NAT rules**: 100, 200, 300... (separate from destination)
- **Destination NAT rules**: 100, 200, 300... (separate from source)
- **Binat rules**: 500, 600, 700... (both source and destination to avoid conflicts)
- Increment by 100 to allow manual rule insertion if needed

### VyOS Sagitta Syntax
All commands use VyOS Sagitta syntax:
```
set nat source rule 100 outbound-interface name 'eth0'
set nat source rule 100 source address '10.0.0.0/8'
set nat source rule 100 translation address 'masquerade'

set nat destination rule 100 inbound-interface name 'eth0'
set nat destination rule 100 protocol 'tcp'
set nat destination rule 100 destination port '443'
set nat destination rule 100 translation address '10.62.0.20'
```

### Type Safety
- All generator methods include explicit type narrowing for mypy strict mode
- Pydantic models validate NAT configuration at parse time
- No type: ignore comments needed

### Unit Test Coverage
- **41 new test cases** covering all NAT types
- Tests for rule numbering, mixed configurations, edge cases
- Protocol handling (tcp, udp, tcp_udp, icmp)
- Port translation scenarios
- Binat with and without descriptions
- Conflict prevention between regular NAT and binat rules

### Test Results
```
============================= 364 passed in 1.20s ==============================
```

All tests passing including:
- Linting (ruff)
- Type checking (mypy strict mode)
- Unit tests (pytest)

## Schema Reference

NAT configuration uses the `NAT_JSON` context variable. Example:

```json
{
  "source": [
    {
      "outbound_interface": "eth0",
      "source_address": "10.0.0.0/8",
      "translation": "masquerade"
    }
  ],
  "destination": [
    {
      "inbound_interface": "eth0",
      "protocol": "tcp",
      "destination_port": 443,
      "translation_address": "10.62.0.20",
      "translation_port": 443
    }
  ],
  "binat": [
    {
      "external_address": "129.244.246.66",
      "internal_address": "10.63.0.101",
      "interface": "eth0",
      "description": "Scoring engine"
    }
  ]
}
```

## Related Issues

Part of the phased vyos-onecontext implementation plan (Phase 5: NAT).

## Checklist

- [x] Implementation follows existing generator patterns
- [x] Comprehensive unit tests added (41 test cases)
- [x] All linting checks pass (ruff)
- [x] Type checking passes (mypy strict mode)
- [x] All 364 tests passing
- [x] Documentation in docstrings and comments
- [x] VyOS Sagitta syntax verified

Generated with Claude Code (Sonnet 4.5)

### CODE REVIEW

File: src/vyos_onecontext/generators/nat.py [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+
+            # Protocol (optional - if not specified, matches all protocols)
+            if rule.protocol:
+                # Handle tcp_udp specially - need to set both protocols
+                if rule.protocol == "tcp_udp":
+                    commands.append(f"set nat destination rule {rule_num} protocol 'tcp_udp'")
+                else:
+                    commands.append(
+                        f"set nat destination rule {rule_num} protocol '{rule.protocol}'"
+                    )
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/nat.py]: Fixed in commit 979258d. Removed the redundant tcp_udp special handling - VyOS accepts it as a protocol value directly.

(AI-generated via Claude Code)

File: tests/test_generators.py [hunk 2]
```diff
@@ -1310,7 +1315,7 @@ def test_generate_gateway_fallback_to_later_interface(self):
         gen = VrfGenerator(interfaces)
         commands = gen.generate()
 
-        # eth1 should provide the gateway since eth0's is invalid
+        # eth1 should provide the gateway since eth0s is invalid
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_generators.py]: Fixed in commit 979258d. Changed eth0s to eth0's (possessive form).

(AI-generated via Claude Code)

File: tests/test_generators.py [hunk 3]
```diff
@@ -2265,3 +2270,410 @@ def test_generate_config_command_order_with_dhcp(self):
...
+                    translation="masquerade",
+                    description="NAT for internal network",
+                )
+            ]
+        )
+        gen = NatGenerator(nat)
+        commands = gen.generate()
+
+        assert len(commands) == 4
+        assert "set nat source rule 100 description NAT for internal network" in commands
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_generators.py]: Fixed in commit 979258d. Updated test assertion to expect quoted description.

(AI-generated via Claude Code)

File: tests/test_generators.py [hunk 4]
```diff
@@ -2265,3 +2270,410 @@ def test_generate_config_command_order_with_dhcp(self):
...
+                    description="SSH to jump host",
+                )
+            ]
+        )
+        gen = NatGenerator(nat)
+        commands = gen.generate()
+
+        assert "set nat destination rule 100 destination port 2222" in commands
+        assert "set nat destination rule 100 translation port 22" in commands
+        assert "set nat destination rule 100 description SSH to jump host" in commands
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_generators.py]: Fixed in commit 979258d. Updated test assertion to expect quoted description.

(AI-generated via Claude Code)

File: tests/test_generators.py [hunk 5]
```diff
@@ -2265,3 +2270,410 @@ def test_generate_config_command_order_with_dhcp(self):
...
+                    description="Scoring engine",
+                )
+            ]
+        )
+        gen = NatGenerator(nat)
+        commands = gen.generate()
+
+        # Should have description on both rules
+        assert "set nat destination rule 500 description Scoring engine" in commands
+        assert "set nat source rule 500 description Scoring engine" in commands
```
[the remark this answers is not in the document]
Comment (duplico) [reply tests/test_generators.py]: Fixed in commit 979258d. Updated test assertions to expect quoted descriptions on both binat rules.

(AI-generated via Claude Code)

File: src/vyos_onecontext/generators/nat.py [hunk 6]
```diff
@@ -0,0 +1,228 @@
...
+            if rule.translation == "masquerade":
+                commands.append(f"set nat source rule {rule_num} translation address masquerade")
+            elif rule.translation_address:
+                commands.append(
+                    f"set nat source rule {rule_num} translation address {rule.translation_address}"
+                )
+
+            # Description (optional)
+            if rule.description:
+                commands.append(f"set nat source rule {rule_num} description {rule.description}")
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/nat.py]: Fixed in commit 979258d. Added single quotes around description value to ensure proper shlex.split() parsing when the wrapper processes commands.

(AI-generated via Claude Code)

File: src/vyos_onecontext/generators/nat.py [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+            # Translation port (optional - if not specified, port is preserved)
+            if rule.translation_port is not None:
+                commands.append(
+                    f"set nat destination rule {rule_num} translation port {rule.translation_port}"
+                )
+
+            # Description (optional)
+            if rule.description:
+                commands.append(
+                    f"set nat destination rule {rule_num} description {rule.description}"
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/nat.py]: Fixed in commit 979258d. Added single quotes around description value to ensure proper shlex.split() parsing when the wrapper processes commands.

(AI-generated via Claude Code)

File: src/vyos_onecontext/generators/nat.py [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            commands.append(
+                f"set nat source rule {rule_num} translation address {rule.external_address}"
+            )
+
+            # Description (applied to both rules if provided)
+            if rule.description:
+                commands.append(
+                    f"set nat destination rule {rule_num} description {rule.description}"
+                )
+                commands.append(f"set nat source rule {rule_num} description {rule.description}")
```
[the remark this answers is not in the document]
Comment (duplico) [reply src/vyos_onecontext/generators/nat.py]: Fixed in commit 979258d. Added single quotes around description values on both binat rules (source and destination) to ensure proper shlex.split() parsing when the wrapper processes commands.

(AI-generated via Claude Code)
